crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,266 @@
+                    caller.data_mut().result_len = len;
+                },
+            )
+            .unwrap()
+            .func_wrap(
+                "typst_env",
+                "wasm_minimal_protocol_write_args_to_buffer",
+                move |mut caller: Caller<PersistentData>, ptr: u32| {
+                    let memory =
+                        caller.get_export("memory").unwrap().into_memory().unwrap();
+                    let buffer = std::mem::take(&mut caller.data_mut().arg_buffer);
+                    memory.write(&mut caller, ptr as _, &buffer).unwrap();
+                    caller.data_mut().arg_buffer = buffer;
+                },
+            )
+            .unwrap()
+            .instantiate(&mut store, &module)
+            .map_err(|e| format!("{e}"))?
+            .start(&mut store)
+            .map_err(|e| format!("{e}"))?;
+
+        let mut free_function = None;
+        let functions = instance
+            .exports(&store)
+            .filter_map(|e| {
+                let name = e.name().to_owned();
+
+                e.into_func().map(|func| {
+                    if name == "wasm_minimal_protocol_free_byte_buffer" {
+                        free_function = Some(func);
+                    }
+                    (name, func)
+                })
+            })
+            .collect::<Vec<_>>();
+        let free_function = free_function.ok_or(EcoString::from("Module didn't export a free function."))?;
+        let memory =
+            instance.get_export(&store, "memory").unwrap().into_memory().unwrap();
+        Ok(Plugin(Arc::new(Repr {
+            store: Mutex::new(store),
+            memory,
+            free_function,
+            functions,
+            bytes: bytes.as_ref().into(),
+        })))
+    }
+
+    fn store(&self) -> MutexGuard<'_, Store> {
+        self.0.store.lock().unwrap()
+    }
+
+    /// Call a function defined in the plugin under `function_name`.
+    ///
+    /// This will eat the number of argument it needs of type Bytes.
+    ///
+    /// # Errors
+    /// - if the plugin doesn't contain the function
+    /// - if the number of argument isn't correct
+    pub fn call(&self, function_name: &str, args: &mut Args) -> SourceResult<Value> {
+        let span = args.span;
+        let ty = self
+            .get_function(function_name)
+            .ok_or("plugin doesn't have the method: {function}")
+            .at(span)?
+            .ty(self.store().as_context());
+        let arg_count = ty.params().len();
+        let mut byte_args = vec![];
+        for k in 0..arg_count {
+            let arg = args
+                .eat::<Value>()?
+                .ok_or(format!("plugin methods takes {arg_count} args, {k} provided"))
+                .at(span)?
+                .cast::<Bytes>()
+                .at(span)?;
+            byte_args.push(arg);
+        }
+        let byte_args = byte_args.iter().map(|b| b.as_slice()).collect::<Vec<_>>();
```

Comment:
No need to collect here

Reply:
I think it's needed because of the line just below!

Reply:
Bonus: it should be possible to avoid allocating a buffer for `byte_args` entirely! Something like
```rust
let byte_args = (0..arg_count).map(|k| -> SourceResult<Bytes> {
    args.eat::<Value>()?
        .ok_or(format!("plugin methods takes {arg_count} args, {k} provided"))
        .at(span)?
        .cast::<Bytes>()
        .at(span)
});
```
And
```rust
fn call_inner(&self, function: &str, args: impl IntoIterator<Item = SourceResult<Bytes>>) -> SourceResult<ReturnedData> {
    // ...
}
```